devito/propagator.py

```diff
@@ -281,56 +283,75 @@ def generate_space_loops(self, loop_body):
             if ivdep and len(self.space_dims) > 1:
                 loop_body = cgen.Block([self.compiler.pragma_ivdep] + [loop_body])
             ivdep = False
-        return loop_body
+        return [loop_body]  # returns body as a list
 
     def generate_space_loops_blocking(self, loop_body):
-        """Generate cgen block for a cache blocking space loop
-
+        """Generate list<cgen.For> for a cache blocking space loop
         :param loop_body: Statement representing the loop body
-        :returns: :class:`cgen.Block` representing the loop
+        :returns: :list<cgen.For> a list of for loops
         """
         ivdep = True
         remainder = False
         orig_loop_body = loop_body
+        remainder_loop = None
+
         for spc_var, block_size in reversed(zip(list(self.space_dims), self.block_sizes)):
             dim_var = str(self._var_map[spc_var])
             block_var = dim_var + "b"
             loop_limits = self._space_loop_limits[spc_var]
-            loop_body = cgen.For(cgen.InlineInitializer(cgen.Value("int", dim_var),
-                                                        block_var),
-                                 dim_var + "<" + block_var+"+"+str(block_size), dim_var + "++", loop_body)
+
+            lower_limit_str = block_var
+            upper_limit_str = block_var+"+"+str(block_size)
+
+            if ivdep and not self.cache_block_inner:
+                lower_limit_str = str(loop_limits[0])
+                upper_limit_str = str(loop_limits[1])
+
+            loop_body = cgen.For(cgen.InlineInitializer(cgen.Value("int", dim_var), lower_limit_str),
+                                 dim_var + "<" + upper_limit_str, dim_var + "++", loop_body)
+
             if ivdep and len(self.space_dims) > 1:
                 loop_body = cgen.Block([self.compiler.pragma_ivdep] + [loop_body])
             ivdep = False
 
+        ivdep = True
```

Comment:
Please rename this flag to something more indicative of its use: `innermost_dim` or something.